src/libsyntax/ext/derive.rs

```diff
@@ -0,0 +1,250 @@
+// Copyright 2012-2017 The Rust Project Developers. See the COPYRIGHT
+// file at the top-level directory of this distribution and at
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+use ast::{self, NestedMetaItem};
+use ext::base::{ExtCtxt, SyntaxExtension};
+use ext::build::AstBuilder;
+use feature_gate;
+use symbol::Symbol;
+use syntax_pos::Span;
+use codemap;
+
+pub fn is_derive_attr(attr: &ast::Attribute) -> bool {
+    let res = attr.name() == Symbol::intern("derive");
+    res
+}
+
+pub fn derive_attr_trait<'a>(cx: &mut ExtCtxt, attr: &'a ast::Attribute)
+                             -> Option<&'a NestedMetaItem> {
+    if !is_derive_attr(attr) {
+        return None;
+    }
+    if attr.value_str().is_some() {
+        cx.span_err(attr.span, "unexpected value in `derive`");
+        return None;
+    }
+
+    let traits = attr.meta_item_list().unwrap();
+
+    if traits.is_empty() {
+        cx.span_warn(attr.span, "empty trait list in `derive`");
+        return None;
+    }
+
+    return traits.get(0);
+}
+
+pub fn verify_derive_attrs(cx: &mut ExtCtxt, attrs: &mut Vec<ast::Attribute>) {
+    for i in 0..attrs.len() {
+        if !is_derive_attr(&attrs[i]) {
+            continue;
+        }
+
+        let span = attrs[i].span;
+
+        if attrs[i].value_str().is_some() {
+            cx.span_err(span, "unexpected value in `derive`");
+        }
+
+        let traits = attrs[i].meta_item_list().unwrap_or(&[]).to_owned();
+
+        if traits.is_empty() {
+            cx.span_warn(span, "empty trait list in `derive`");
+            continue;
+        }
+        for titem in traits {
+            if titem.word().is_none() {
+                cx.span_err(titem.span, "malformed `derive` entry");
+            }
+        }
+    }
+}
+
+pub fn get_derive_attr(cx: &mut ExtCtxt, attrs: &mut Vec<ast::Attribute>,
+                       is_derive_type: fn(&mut ExtCtxt, &NestedMetaItem) -> bool)
```

Comment:
I think it would be a little cleaner if this were an enum (three variants) and `is_derive_type(...)` were a `match`.